crates/wit-parser/src/ast.rs

```diff
@@ -82,60 +82,162 @@ impl<'a> DeclList<'a> {
         }
     }
 
-    fn parse_implicit_package_items(
+    fn parse_main_package_items(
         tokens: &mut Tokenizer<'a>,
         mut docs: Docs<'a>,
-    ) -> Result<DeclList<'a>> {
+        mut nested: Vec<NestedPackage<'a>>,
+        main_pkg_decl: Option<&PackageName>,
+    ) -> Result<(DeclList<'a>, Vec<NestedPackage<'a>>)> {
         let mut items = Vec::new();
         while tokens.clone().next()?.is_some() {
-            items.push(AstItem::parse(tokens, docs)?);
-            docs = parse_docs(tokens)?;
+            if tokens.eat(Token::Package)? {
+                let package_id = PackageName::parse(tokens, &mut docs)?;
+                if let Some(main) = main_pkg_decl {
+                    if main.package_name() == package_id.package_name() {
+                        bail!(
+                            "Nested package name already used by main package declaration: {}",
+                            main.package_name()
+                        );
+                    }
+                }
+                if tokens.eat(Token::LeftBrace)? {
+                    let redundant = nested
+                        .iter()
+                        .find(|p| p.package_id.package_name() == package_id.package_name());
+
+                    if redundant.is_some() {
+                        bail!(
+                            "colliding nested package names, multiple packages named `{}`",
+                            package_id.package_name()
+                        )
+                    }
+                    nested.push(NestedPackage {
+                        package_id,
+                        decl_list: DeclList::parse_nested_package_items(tokens)?,
+                    });
+                } else if tokens.eat(Token::Semicolon)? {
+                    if let (parsed, Some(main)) = (&package_id, main_pkg_decl) {
+                        let parsed_name = parsed.package_name();
+                        let main_name = main.package_name();
+                        bail!(
+                            "Multiple main package declarations in one file: {} and {}",
+                            parsed_name,
+                            main_name
+                        );
+                    }
+                }
+            } else {
+                items.push(AstItem::parse(tokens, docs.clone())?);
+                docs = parse_docs(tokens)?;
+            }
         }
-        Ok(DeclList { items })
+        Ok((DeclList { items }, nested))
     }
 }
 
-impl<'a> Ast<'a> {
-    fn parse(tokens: &mut Tokenizer<'a>) -> Result<Self> {
+impl<'a> SourceAst<'a> {
+    fn parse(
```

Comment:
The previous split of parsing implicit/explicit items doesn't quite make sense any more with this and I find that the structure here is a bit confusing. Could this instead be modelled as:

1. Parse the `package foo:bar;` at the top of the file, this is unconditionally required to come first.
2. Parse one-item-at-a-time after that which is either a nested package or an ast item.
3. When empty, return

I think that would help simplify a good chunk of the parsing here and in `parse_main_package_items` perhaps? For example an `AstItem` could be expanded to hold `NestedPackage(...)` or something like that. That would AST-wise allow arbitrarily nested packages but the AST validator could reject that saying that nested packages are only supported at the top level.